src/miner.cpp

```diff
@@ -134,41 +166,40 @@ class AddOutputsToCoinbaseTxAndSign : public boost::static_visitor<>
-        auto value = GetBlockSubsidy(nHeight, chainparams.GetConsensus());
-
-        if ((nHeight > 0) && (nHeight <= chainparams.GetConsensus().GetLastFoundersRewardBlockHeight(nHeight))) {
-            // Founders reward is 20% of the block subsidy
-            auto vFoundersReward = value / 5;
-            // Take some reward away from us
-            value -= vFoundersReward;
-            // And give it to the founders
-            mtx.vout.push_back(CTxOut(vFoundersReward, chainparams.GetFoundersRewardScriptAtHeight(nHeight)));
+    CAmount SetFoundersRewardAndGetMinerValue(void* ctx) const {
+        auto miner_reward = GetBlockSubsidy(nHeight, chainparams.GetConsensus());
+
+        if (nHeight > 0) {
+            if (!chainparams.GetConsensus().NetworkUpgradeActive(nHeight, Consensus::UPGRADE_CANOPY)) {
+                if (nHeight <= chainparams.GetConsensus().GetLastFoundersRewardBlockHeight(nHeight)) {
+                    // Founders reward is 20% of the block subsidy
+                    auto vFoundersReward = miner_reward / 5;
+                    // Take some reward away from us
+                    miner_reward -= vFoundersReward;
+                    // And give it to the founders
+                    mtx.vout.push_back(CTxOut(vFoundersReward, chainparams.GetFoundersRewardScriptAtHeight(nHeight)));
+                }
+            } else {
+                auto requiredShares = Consensus::GetActiveFundingStreamShares(
+                    nHeight,
+                    miner_reward,
+                    chainparams.GetConsensus());
+
+                BOOST_FOREACH(Consensus::FundingStreamShare share, requiredShares) {
+                    miner_reward -= share.second;
+                    bool added = boost::apply_visitor(AddFundingStreamShareToTx(mtx, ctx, share.second), share.first);
+                    if (!added) {
+                        librustzcash_sapling_proving_ctx_free(ctx);
+                        throw new std::runtime_error("Failed to add funding stream output.");
+                    }
+                }
+            }
         }
 
-        return value + nFees;
+        return miner_reward + nFees;
     }
 
-    void operator()(const InvalidMinerAddress &invalid) const {}
-
-    // Create shielded output
-    void operator()(const libzcash::SaplingPaymentAddress &pa) const {
-        auto value = SetFoundersRewardAndGetMinerValue();
-        mtx.valueBalance = -value;
-
-        uint256 ovk;
-        auto note = libzcash::SaplingNote(pa, value);
-        auto output = OutputDescriptionInfo(ovk, note, {{0xF6}});
-
-        auto ctx = librustzcash_sapling_proving_ctx_init();
-
-        auto odesc = output.Build(ctx);
-        if (!odesc) {
-            librustzcash_sapling_proving_ctx_free(ctx);
-            throw new std::runtime_error("Failed to create shielded output for miner");
-        }
-        mtx.vShieldedOutput.push_back(odesc.get());
-
+    void ComputeBindingSig(void* ctx) const {
```

Comment:
This isn't affected by this PR, but the call to `librustzcash_sapling_binding_sig` in this function doesn't check if the result is `true` and throw an exception otherwise. I think that's a bug but feel free to just file an issue.